addons/mail/controllers/mail.py

```diff
@@ -95,16 +96,25 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
             else:
                 record_action = record_sudo.get_access_action(access_uid=uid)
         else:
-            record_action = record_sudo.get_access_action()
-            if record_action['type'] == 'ir.actions.act_url' and record_action.get('target_type') != 'public':
-                return cls._redirect_to_messaging()
+            act = record_action = record_sudo.get_access_action()
+            # for an unlogged user, anything other than a public url should
+            # bounce through the login page
+            if (act['type'], act.get('target_type')) != ('ir.actions.act_url', 'public'):
+                return werkzeug.utils.redirect(
+                    '/web/login?' + url_encode({
+                        'redirect': '/mail/view?' + url_encode({
```

Comment:
Technically I think we also support message_id and access_token (and pid, hash, auth_signup_token and auth_login in portal). Don't remember if everything is useful and all the flows, woud be interesting to dive in it someday, but we could at least support them and see afterwards :D (or maybe we give kwargs as we receive them and basta ?).